File: sda-tp2/C/src/main.c
#include<stdio.h>
#include <time.h>
#include<stdlib.h>
#include "structures/arraylist.h"
#include "analysis/analyzer.h"


int main(int argc, char ** argv){
  int i;
  // Tableau dynamique.
  arraylist_t * a = arraylist_create();
  // Analyse du temps pris par les opérations.
  analyzer_t * time_analysis = analyzer_create();
  // Analyse du nombre de copies faites par les opérations.
  analyzer_t * copy_analysis = analyzer_create();
  // Analyse de l'espace mémoire inutilisé.
  analyzer_t * memory_analysis = analyzer_create(); 
  // Mesure de la durée d'une opération.
  struct timespec before, after;
  clockid_t clk_id = CLOCK_REALTIME;
  // utilisé comme booléen pour savoir si une allocation a été effectuée.
  char memory_allocation; 
  srand(0);
  
  for(i = 0; i < 1000000 ; i++){
    if(rand()%2){
      // Ajout d'un élément et mesure du temps pris par l'opération.    
      clock_gettime(clk_id, &before);
      memory_allocation = arraylist_append(a, i);
      clock_gettime(clk_id, &after);
    }else{
      // Suppression du dernier élément et mesure du temps pris par l'opération.    
      timespec_get(&before, TIME_UTC);
      memory_allocation = arraylist_pop_back(a);
      timespec_get(&after, TIME_UTC);
    }
    // Enregistrement du temps pris par l'opération
    analyzer_append(time_analysis, after.tv_nsec - before.tv_nsec);
    // Enregistrement du nombre de copies efféctuées par l'opération.
    // S'il y a eu réallocation de mémoire, il a fallu recopier tout le tableau.
    analyzer_append(copy_analysis, (memory_allocation)? arraylist_size(a):1 );
    // Enregistrement de l'espace mémoire non-utilisé.
    analyzer_append(memory_analysis,arraylist_capacity(a)-arraylist_size(a));

  }

  // Affichage de quelques statistiques sur l'expérience.
  fprintf(stderr, "Total cost: %Lf\n", get_total_cost(time_analysis));
  fprintf(stderr, "Average cost: %Lf\n", get_average_cost(time_analysis));
  fprintf(stderr, "Variance: %Lf\n", get_variance(time_analysis));
  fprintf(stderr, "Standard deviation: %Lf\n", get_standard_deviation(time_analysis));

  // Sauvegarde les données de l'expérience.
  save_values(time_analysis, "../plots/dynamic_array_add_and_delete_sqrt_time_c.plot");
  save_values(copy_analysis, "../plots/dynamic_array_add_and_delete_sqrt_copy_c.plot");
  save_values(memory_analysis, "../plots/dynamic_array_add_and_delete_sqrt_memory_c.plot");

  // Nettoyage de la mémoire avant la sortie du programme
  arraylist_destroy(a);
  analyzer_destroy(time_analysis);
  analyzer_destroy(copy_analysis);
  analyzer_destroy(memory_analysis);
  return EXIT_SUCCESS;
}

File: sda-tp2/Java/Main.java
import analysis.Analyzer;
import structures.ArrayList;

import java.util.Random;

public class Main {

	public static void main(String[] args) {
		int i, time_ind = 0, pop_ind = 0;
		// Tableau dynamique.
		ArrayList<Integer> a = new ArrayList<Integer>();
		// Analyse du temps pris par les opérations.
		Analyzer time_analysis = new Analyzer();
		// Analyse du nombre de copies faites par les opérations.
		Analyzer copy_analysis = new Analyzer();
		// Analyse de l'espace mémoire inutilisé.
		Analyzer memory_analysis = new Analyzer();
		long before, after;
		// Booléen permettant de savoir si une allocation a été effectuée.
		boolean memory_allocation;
		Random random_generator = new Random();
		float prob = 0.8f;
		for (i = 0; i < 1000000; i++) {
			if (random_generator.nextFloat() < prob) {
				// Ajout d'un élément et mesure du temps pris par l'opération.
				before = System.nanoTime();
				memory_allocation = a.append(i);
				after = System.nanoTime();

				// Enregistrement du temps pris par l'opération
				time_analysis.append(after - before);
				// Enregistrement du nombre de copies efféctuées par l'opération.
				// S'il y a eu réallocation de mémoire, il a fallu recopier tout le tableau.
				copy_analysis.append((memory_allocation == true) ? a.size() : 1);
				// Enregistrement de l'espace mémoire non-utilisé.
				memory_analysis.append(a.capacity() - a.size());

			} else {
				if (a.size() > 0) {
					before = System.nanoTime();
					memory_allocation = a.pop_back();
					after = System.nanoTime();

					// Enregistrement du temps pris par l'opération
					time_analysis.append(after - before);
					// Enregistrement du nombre de copies efféctuées par l'opération.
					// S'il y a eu réallocation de mémoire, il a fallu recopier tout le tableau.
					copy_analysis.append((memory_allocation == true) ? a.size() : 1);
					// Enregistrement de l'espace mémoire non-utilisé.
					memory_analysis.append(a.capacity() - a.size());

				}
			}

		}

		// Affichage de quelques statistiques sur l'expérience.
		System.err.println("Total cost : " + time_analysis.get_total_cost());
		System.err.println("Average cost : " + time_analysis.get_average_cost());
		System.err.println("Variance :" + time_analysis.get_variance());
		System.err.println("Standard deviation :" + time_analysis.get_standard_deviation());

		// Sauvegarde les données de l'expérience: temps et nombre de copies effectuées
		// par opération.
		time_analysis.save_values("../plots/dynamic_array_time_java_INSDEL_p" + prob + "_3.plot");
		copy_analysis.save_values("../plots/dynamic_array_sqrt_copy_java_INSDEL_p" + prob + "_3.plot");
		memory_analysis.save_values("../plots/dynamic_array_sqrt_memory_java_INSDEL_p" + prob + "_3.plot");
	}
}
